trace/client/src/main/scala/org/http4s/otel4s/middleware/trace/client/AttributeProvider.scala

```diff
@@ -0,0 +1,109 @@
+  * had their bodies replaced with empty `Stream`s (by calling
+  * `.withBodyStream(Stream.empty)`).
+  *
+  * @note Implementations MUST NOT access request or response bodies.
+  */
+trait AttributeProvider { self =>
+
+  /** Provides attributes for a span using the given request.
+    *
+    * It is RECOMMENDED that callers pass `Request`s that have had their bodies
+    * replaced with empty `Stream`s (by calling `.withBodyStream(Stream.empty)`).
+    *
+    * @note Implementation MUST NOT access request body.
+    */
+  def requestAttributes[F[_]](
+      request: Request[F],
+      urlTemplateClassifier: UriTemplateClassifier,
+      urlRedactor: UriRedactor,
+      headersAllowedAsAttributes: Set[CIString],
+  ): Attributes
+
+  /** Provides attributes for a span using the given response.
+    *
+    * It is RECOMMENDED that callers pass `Response`s that have had their bodies
+    * replaced with empty `Stream`s (by calling `.withBodyStream(Stream.empty)`).
+    *
+    * @note Implementation MUST NOT access response body.
+    */
+  def responseAttributes[F[_]](
+      response: Response[F],
+      headersAllowedAsAttributes: Set[CIString],
+  ): Attributes
+
+  /** @return an `AttributeProvider` that provides the attributes from this and
+    *         another `AttributeProvider`
+    */
+  def and(that: AttributeProvider): AttributeProvider =
+    new AttributeProvider {
+      def requestAttributes[F[_]](
+          request: Request[F],
+          urlTemplateClassifier: UriTemplateClassifier,
+          urlRedactor: UriRedactor,
+          headersAllowedAsAttributes: Set[CIString],
+      ): Attributes =
+        self.requestAttributes(
+          request,
+          urlTemplateClassifier,
+          urlRedactor,
+          headersAllowedAsAttributes,
+        ) ++
+          that.requestAttributes(
+            request,
+            urlTemplateClassifier,
+            urlRedactor,
+            headersAllowedAsAttributes,
+          )
+
+      def responseAttributes[F[_]](
+          response: Response[F],
+          headersAllowedAsAttributes: Set[CIString],
+      ): Attributes =
+        self.responseAttributes(response, headersAllowedAsAttributes) ++
+          that.responseAttributes(response, headersAllowedAsAttributes)
+    }
+}
+
+object AttributeProvider {
+
+  /** Provides an `Attribute` containing this middleware's version. */
+  val middlewareVersion: AttributeProvider =
+    new AttributeProvider {
+      def requestAttributes[F[_]](
+          request: Request[F],
+          urlTemplateClassifier: UriTemplateClassifier,
+          urlRedactor: UriRedactor,
+          headersAllowedAsAttributes: Set[AuthScheme],
+      ): Attributes = Attributes(TypedAttributes.middlewareVersion)
+      def responseAttributes[F[_]](
+          response: Response[F],
+          headersAllowedAsAttributes: Set[AuthScheme],
+      ): Attributes = Attributes.empty
+    }
```

Comment:
I'm not sure if this is a reasonable thing to do or not, but it's a separate commit so it will be easy to remove if we decide not